ruma-federation-api/src/transactions/send_transaction_message/edu.rs

```diff
@@ -0,0 +1,294 @@
+//! [PUT /_matrix/federation/v1/send/{txnId}](https://matrix.org/docs/spec/server_server/r0.1.3#put-matrix-federation-v1-send-txnid)
+
+use std::collections::BTreeMap;
+
+use js_int::UInt;
+use ruma_common::{encryption::DeviceKeys, presence::PresenceState};
+use ruma_events::{from_raw_json_value, receipt::Receipt};
+use ruma_identifiers::{DeviceIdBox, EventId, RoomId, UserId};
+use serde::{de, Deserialize, Serialize};
+use serde_json::{value::RawValue as RawJsonValue, Value as JsonValue};
+
+// There is one more edu_type synapse recognizes with the note:
+// FIXME: switch to "m.signing_key_update" when MSC1756 is merged into the
+// spec from "org.matrix.signing_key_update"
+
+/// Type for passing ephemeral data to homeservers.
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "edu_type", content = "content")]
+pub enum Edu {
+    /// An EDU representing presence updates for users of the sending homeserver.
+    #[serde(rename = "m.presence")]
+    Presence(PresenceContent),
+
+    /// An EDU representing receipt updates for users of the sending homeserver.
+    #[serde(rename = "m.receipt")]
+    Receipt(ReceiptContent),
+
+    /// A typing notification EDU for a user in a room.
+    #[serde(rename = "m.typing")]
+    Typing(TypingContent),
+
+    /// An EDU that lets servers push details to each other when one of their users adds
+    /// a new device to their account, required for E2E encryption to correctly target the
+    /// current set of devices for a given user.
+    #[serde(rename = "m.device_list_update")]
+    DeviceListUpdate(DeviceListUpdateContent),
+
+    /// An EDU that lets servers push send events directly to a specific device on a
+    /// remote server - for instance, to maintain an Olm E2E encrypted message channel
+    /// between a local and remote device.
+    #[serde(rename = "m.direct_to_device")]
+    DirectToDevice(DirectDeviceContent),
+
+    #[doc(hidden)]
+    Custom(JsonValue),
+}
+
+#[derive(Debug, Deserialize)]
+struct EduTpeDeHelper {
+    /// The message type field
+    edu_type: String,
+    content: Box<RawJsonValue>,
+}
+
+impl<'de> de::Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduTpeDeHelper { edu_type, ref content } = from_raw_json_value(&json)?;
```

Comment:
I generally avoid `ref`, why would it be needed here? Just to avoid `&` in the branches below?

Reply:
Yeah just so I didn't have to write `&` so much. I'm fine either way.

Reply:
Then let's remove `ref`.